x-pack/plugins/licensing/server/plugin.ts

```diff
@@ -0,0 +1,136 @@
+  private readonly logger: Logger;
+  private readonly config$: Observable<LicensingConfig>;
+  private poller!: Poller<ILicense>;
+
+  constructor(private readonly context: PluginInitializerContext) {
+    this.logger = this.context.logger.get();
+    this.config$ = this.context.config
+      .create<LicensingConfigType | { config: LicensingConfigType }>()
+      .pipe(
+        map(config =>
+          'config' in config
+            ? new LicensingConfig(config.config, this.context.env)
+            : new LicensingConfig(config, this.context.env)
+        )
+      );
+  }
+
+  private hasLicenseInfoChanged(newLicense: any) {
+    const currentLicense = this.poller.subject$.getValue();
+
+    if ((currentLicense && !newLicense) || (newLicense && !currentLicense)) {
+      return true;
+    }
+
+    return (
+      newLicense.type !== currentLicense.type ||
+      newLicense.status !== currentLicense.status ||
+      newLicense.expiry_date_in_millis !== currentLicense.expiryDateInMillis
+    );
+  }
+
+  private async fetchInfo(core: CoreSetup, clusterSource: string, pollingFrequency: number) {
+    this.logger.debug(
+      `Calling [${clusterSource}] Elasticsearch _xpack API. Polling frequency: ${pollingFrequency}`
+    );
+
+    const cluster = await core.elasticsearch.dataClient$.pipe(first()).toPromise();
+
+    try {
+      const response = await cluster.callAsInternalUser('transport.request', {
+        method: 'GET',
+        path: '/_xpack',
+      });
+      const rawLicense = response && response.license;
+      const features = (response && response.features) || {};
+      const licenseInfoChanged = this.hasLicenseInfoChanged(rawLicense);
+
+      if (!licenseInfoChanged) {
+        return { license: false, error: null, features: null };
+      }
+
+      const currentLicense = this.poller.subject$.getValue();
+      const licenseInfo = [
+        'type' in rawLicense && `type: ${rawLicense.type}`,
+        'status' in rawLicense && `status: ${rawLicense.status}`,
+        'expiry_date_in_millis' in rawLicense &&
+          `expiry date: ${moment(rawLicense.expiry_date_in_millis, 'x').format()}`,
+      ]
+        .filter(Boolean)
+        .join(' | ');
+
+      this.logger.info(
+        `Imported ${currentLicense ? 'changed ' : ''}license information` +
+          ` from Elasticsearch for the [${clusterSource}] cluster: ${licenseInfo}`
+      );
+
+      return { license: rawLicense, error: null, features };
+    } catch (err) {
+      this.logger.warn(
+        `License information could not be obtained from Elasticsearch` +
+          ` for the [${clusterSource}] cluster. ${err}`
+      );
+
+      return { license: null, error: err, features: {} };
+    }
+  }
+
+  private create({ clusterSource, pollingFrequency }: LicensingConfig, core: CoreSetup) {
+    this.poller = new Poller<ILicense>(
+      pollingFrequency,
+      new License(null, {}, null, clusterSource),
```

Comment:
1. signature a bit criptic. does it make sense to pass as object? 
```js
new License({
  license: null,
  features: {},
  error: null,
  clusterSource
});
```
2. What is the point to pass an error?

Reply:
1. Resolved in next PR.
2. Used by `reasonUnavailable`.